File: CarRentalFrontend/src/app/core/services/carinventory.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { environment } from 'src/environments/environment';
import { Car } from '../Models/car';
import { BehaviorSubject } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class CarinventoryService {
  OrderDetails = new BehaviorSubject({});
  constructor(private http: HttpClient) {

  }

  AddCar(car: Car) {
    let url = environment.api + 'carinventory/addcar';
    return this.http.post(url, car);
  }

  DeleteCar(vehicleId: string) {
    let url = environment.api + `carinventory/deletecar/${vehicleId}`;
    return this.http.delete(url);
  }

  UpdateCar(car: Car) {
    let url = environment.api + 'carinventory/updatecar';
    return this.http.put(url, car);
  }

  GetCarById(id: string) {
    let url = environment.api + `carinventory/GetCarDetail/${id}`;
    return this.http.get<Car>(url);
  }

  GetAllCar(maker: string, model: string, minPrice: string, maxPrice: string,startDate:string,endDate:string) {
    let url = environment.api + 'carinventory/getallavailablecar?';

    if (minPrice) {
      url = url + `minprice=${minPrice}&`;
    } else {
      url = url + `minprice=${0}&`;
    }

    if (maxPrice) {
      url = url + `maxprice=${maxPrice}&`;
    } else {
      url = url + `maxprice=${10000000}&`;
    }

    url=url+`startDate=${startDate}&`;
    url=url+`endDate=${endDate}&`;
    
    if (maker) {
      url = url + `maker=${maker}&`;
    }


    if (model) {
      url = url + `model=${model}&`;
    }

    url = url.substring(0, (url.length - 1));
    return this.http.get<[Car]>(url);
  }

  GetAllCars() {
    let url=environment.api+'carinventory/allcars';
    return this.http.get<[Car]>(url);

  }


  IsPossibleToBookCar (vehicleId:string,startDate:string,endDate:string) {
    let url=environment.api+`carinventory/IsPossibleToBookCar?vehicleId=${vehicleId}&startDate=${startDate}&endDate=${endDate}`;
    return this.http.get(url);
  }
}

File: Business Layer/Services/CarInventoryService/CarInventoryService.cs
using Data_Access_Layer.Entities;
using Data_Access_Layer.Repositories.CarRepository;
using Data_Access_Layer.Repositories.RentalRepository;
using Data_Access_Layer.UnitOfWork;

namespace Business_Layer.Services.CarService
{
    public class CarInventoryService : ICarInventoryService
    {
        private readonly ICarInventoryRepository _carInventoryRepository;
        private readonly IUnitOfWork _unitOfWork;
        private readonly IRentalRepository _rentalRepository;
        public CarInventoryService(ICarInventoryRepository carRepository, IUnitOfWork unitOfWork, IRentalRepository rentalRepository)
        {
            _carInventoryRepository = carRepository;
            _unitOfWork = unitOfWork;
            _rentalRepository = rentalRepository;
        }

        public async Task<bool> AddCar(Car car)
        {
            car.VehicleId = Guid.NewGuid();
            await _carInventoryRepository.AddAsync(car);
            await _unitOfWork.SaveAsyc();
            return true;
        }

        public async Task<bool> DeleteCar(Guid vehicleId)
        {
            await _carInventoryRepository.Delete((car) => car.VehicleId == vehicleId);
            await _unitOfWork.SaveAsyc();
            return true;
        }

        public async Task<IEnumerable<Car>> GetAllAvailableCar(string? maker, string? model, int minprice, int maxprice,string startDate,string endDate)
        {
            if (maker == null)
            {
                maker = "";
            }

            if (model == null) { model = ""; }

            var vehicles = (await _carInventoryRepository.Filter((car) => car.RentalPrice >= minprice && car.RentalPrice <= maxprice && car.Maker.Contains(maker) && car.Model.Contains(model))).ToList();
            var agreements = (await _rentalRepository.GetAll()).ToList();
            var result = new List<Car>();
            var startDateA = DateTime.Parse(startDate);
            var endDateA = DateTime.Parse(endDate);

            for (var i = 0; i < vehicles.Count(); i++)
            {
                bool isAvailable = true;
                for (var j = 0; j < agreements.Count(); j++)
                {
                    if (vehicles[i].VehicleId == agreements[j].VehicleId)
                    {

                        var startDateB = DateTime.Parse(agreements[j].StartDate);
                        var endDateB = DateTime.Parse(agreements[j].EndDate);

                        var isOverLapping = startDateA <= endDateB && endDateA >= startDateB;

                        if (isOverLapping == true)
                        {
                            isAvailable = false;
                        }
                    }
                }

                if (isAvailable)
                {
                    result.Add(vehicles[i]);
                }
            }

            return result;
        }

        public async Task<IEnumerable<Car>> GetAllCars()
        {
            return await _carInventoryRepository.GetAll();
        }

        public async Task<Car?> GetCarById(Guid vehicleId)
        {
            return await _carInventoryRepository.Find((car) => car.VehicleId == vehicleId);
        }

        public async Task<bool> IsPossibleToBookCar(Guid carId, string SDate, string EDate)
        {

            var startDateA= DateTime.Parse(SDate);
            var endDateA= DateTime.Parse(EDate);
            bool isAvailable = true;
            var agreements = (await _rentalRepository.GetAll()).ToList();
            for (var j = 0; j < agreements.Count(); j++)
            {
                if (carId == agreements[j].VehicleId)
                {

                    var startDateB = DateTime.Parse(agreements[j].StartDate);
                    var endDateB = DateTime.Parse(agreements[j].EndDate);

                    var isOverLapping = startDateA <= endDateB && endDateA >= startDateB;

                    if (isOverLapping == true)
                    {
                        isAvailable = false;
                    }
                }
            }

            return isAvailable;
        }

        public async Task<bool> UpdateCar(Car car)
        {
            _carInventoryRepository.Update(car);
            await _unitOfWork.SaveAsyc();
            return true;
        }
    }
}
